internal/stack/serverless.go

```diff
@@ -0,0 +1,462 @@
+
+	configRegion      = "stack.serverless.region"
+	configProjectType = "stack.serverless.type"
+
+	defaultRegion      = "aws-eu-west-1"
+	defaultProjectType = "observability"
+)
+
+var (
+	errProjectNotExist = errors.New("project does not exist")
+)
+
+type serverlessProvider struct {
+	profile *profile.Profile
+	client  *serverless.Client
+}
+
+type projectSettings struct {
+	Name   string
+	Region string
+	Type   string
+
+	StackVersion string
+}
+
+func (sp *serverlessProvider) createProject(settings projectSettings, options Options, conf Config) (Config, error) {
+	project, err := sp.client.CreateProject(settings.Name, settings.Region, settings.Type)
+	if err != nil {
+		return Config{}, fmt.Errorf("failed to create %s project %s in %s: %w", settings.Type, settings.Name, settings.Region, err)
+	}
+	// project, _ := sp.currentProject(conf)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*30)
+	defer cancel()
+	if err := sp.client.EnsureEndpoints(ctx, project); err != nil {
+		return Config{}, fmt.Errorf("failed to ensure endpoints have been provisioned properly: %w", err)
+	}
+
+	var config Config
+	config.Provider = ProviderServerless
+	config.Parameters = map[string]string{
+		paramServerlessProjectID:   project.ID,
+		paramServerlessProjectType: project.Type,
+	}
+
+	config.ElasticsearchHost = project.Endpoints.Elasticsearch
+	config.KibanaHost = project.Endpoints.Kibana
+	config.ElasticsearchUsername = project.Credentials.Username
+	config.ElasticsearchPassword = project.Credentials.Password
+
+	// Store config now in case fails initialization or other requests,
+	// so it can be destroyed later
+	err = storeConfig(sp.profile, config)
+	if err != nil {
+		return Config{}, fmt.Errorf("failed to store config: %w", err)
+	}
+
+	logger.Debug("Waiting for creation plan to be completed")
+	err = sp.client.EnsureProjectInitialized(ctx, project)
+	if err != nil {
+		return Config{}, fmt.Errorf("project not initialized: %w", err)
+	}
+
+	project, err = sp.createClients(project)
+	if err != nil {
+		return Config{}, fmt.Errorf("failed to create project client")
+	}
+
+	config.Parameters[paramServerlessFleetURL], err = project.DefaultFleetServerURL()
+	if err != nil {
+		return Config{}, fmt.Errorf("failed to get fleet URL: %w", err)
+	}
+	project.Endpoints.Fleet = config.Parameters[paramServerlessFleetURL]
+
+	printUserConfig(options.Printer, config)
+
+	// update config with latest updates (e.g. fleet server url)
+	err = storeConfig(sp.profile, config)
+	if err != nil {
+		return Config{}, fmt.Errorf("failed to store config: %w", err)
+	}
+
+	err = project.EnsureHealthy(ctx)
+	if err != nil {
+		return Config{}, fmt.Errorf("not all services are healthy: %w", err)
+	}
+
+	return config, nil
+}
+
+func (sp *serverlessProvider) createClients(project *serverless.Project) (*serverless.Project, error) {
+	var err error
+	project.ElasticsearchClient, err = NewElasticsearchClient(
```

Comment:
Nit. I would be probably better to keep the clients in the provider, and pass it to the project methods when/if needed. Maybe the healthcheckers could be also here in the provider.

Reply:
I've moved clients to the serverlessProvider struct. So they can be passed to the right methods.